Updated binary encoding details in What's New section (#233)
amzn/ion-docs: src/whatsnew.adoc

=== Motivation

Ion 1.1 has been designed to address some of the trade-offs in Ion 1.0 to make it suitable for a wider range of

> applications.  Ion 1.1 now makes length prefixing of containers optional, and makes the interning of symbolic tokens

Author: 🗺️ Only containers have a delimited form. (When the dust settles on built-in operations, the text and lob types probably will too, though not quite as first-class)

IMPORTANT: *_Discussion_*: Is this statement too weak? Specifically, should we be attempting to "fill in the holes" in
the Ion data model around system values?  Should we require that Ion 1.1 implementations _produce_ Ion 1.0 data?

> Ion 1.1 is *not* required to preserve Ion 1.0 binary encodings in Ion 1.1 encoding contexts (i.e., the type codes and

Author: 🗺️ Subjectively, I thought pluralizing this aligned better with the `(i.e. ...)` that follows.

lower-level encodings are not preserved in the new version).  The Ion Version Marker (IVM) is used to denote the
different versions of the syntax.  Ion 1.1 does retain text compatibility with Ion 1.0 in that the changes are a strict
superset of the grammar, however due to the updated system symbol table, symbol IDs referred to using the `$n` syntax
for symbols beyond the 1.0 system symbol table are not compatible.

[[sec:whatsnew-text]]
=== Text Syntax Changes

Ion 1.1 text *must* use the `$ion_1_1` version marker at the top-level of the data stream or document.

> The only syntax change for the text format is the introduction of *encoding expression* (*E-expression*) syntax, which

Author: 🗺️ Adding "introduction of" caused the whole paragraph to need reformatting. Further down the PR, I started following [the AsciiDoc best practice of putting each source sentence on its own line](https://asciidoctor.org/docs/asciidoc-recommended-practices/#one-sentence-per-line) to mitigate this.

allows for the invocation of macros in the data stream. This syntax is grammatically similar to S-expressions, except that
these expressions are opened with `(:` and closed with `)`.  For example, `(:a 1 2)` would expand the macro named `a` with the
arguments `1` and `2`. See the <<sec:whatsnew-eexp, Macros, Templates, and Encoding-Expressions>> section for details.

This syntax is allowed anywhere an Ion value is allowed:

.Figure {counter:figure-number}. E-expression Examples
[source,plain,%unbreakable]
----
// At the top level
(:foo 1 2)

// Nested in a list
[1, 2, (:bar 3 4)]

// Nested in an S-expression
(cons a (:baz b))

> // Nested in a struct

Author: 🗺️ Made the examples a bit more explicit

IMPORTANT: *_Discussion_*: Should we call this something else (e.g., _non-interned_)?

> In binary Ion 1.0, symbol values, field names, and annotations are required to be encoded using a symbol ID in the local

Author: 🗺️ Another word addition that had rippling formatting effects.

In Ion 1.0, all data is length prefixed.  While this is good for optimizing the reading of data, it requires an Ion
encoder to buffer any data in memory to calculate the data's length.  Ion 1.1 introduces optional binary syntax to allow
containers to be encoded with an end marker.

> ==== Low-level Encoding Changes

Author: 🗺️ This section has been mostly rewritten. The new text links to the relevant subsections of the now-available binary spec.

link:binary-encoding.adoc#flexuint[`FlexUInt`] and link:binary-encoding.adoc#flexint[`FlexInt`].

link:binary-encoding.adoc#flexuint[`FlexUInt`] and link:binary-encoding.adoc#flexint[`FlexInt`] use little-endian byte
order, avoiding the need for reordering on x86 architectures.
Rather than using a bit in each byte to indicate the width of the encoding, `FlexUInt` and `FlexInt` front-load the
continuation bits.
In most cases, this means that these bits all fit in the first byte of the representation, allowing a
reader to determine the complete size of the encoding without having to inspect each byte individually.
Finally, `FlexInt` does not use a separate bit to indicate its value's sign. Instead, it uses two's complement
representation, allowing it to share much of the same structure and parsing logic as its unsigned counterpart.
Benchmarks have shown that in aggregate, these encoding changes are between 1.25 and 3x faster than Ion 1.0's
`VarUInt` and `VarInt` encodings depending on the host architecture.

Ion 1.1 supplants link:https://amazon-ion.github.io/ion-docs/docs/binary.html#uint-and-int-fields[Ion 1.0's `Int`

> encoding primitive] with a new encoding called link:binary-encoding.adoc#fixedint[`FixedInt`], which uses two's

Reviewer: Have we also renamed `UInt` to `FixedUInt`, and should we mention it here?

Author: > Have we also renamed `UInt` to `FixedUInt`

It has, yes.

> and should we mention it here?

I only omitted it because the encoding is the same as `UInt` so there wasn't much to say. I've added a sentence making that explicit.

complement notation instead of sign-and-magnitude.

A new primitive encoding type, link:binary-encoding.adoc#flexsym[`FlexSym`], has been introduced to flexibly encode
symbol IDs and symbolic tokens with inline text.

==== Type Encoding Changes

All Ion types use the new low-level encodings as specified in the previous section.  Many of the opcodes used in Ion 1.0
have been re-organized primarily to make E-expressions compact.

Typed `null` values are now link:binary-encoding.adoc#nulls[encoded in two bytes using the `0xEB` opcode].

Lists and S-expressions have link:binary-encoding.adoc#lists[two encodings]: a length-prefixed encoding and a new
delimited form that ends with
the `0xF0` opcode.

Struct values have link:binary-encoding.adoc#structs[three encodings]: a length-prefixed encoding which uses symbol IDs
for its field names, a length-prefixed encoding which uses `FlexSym` for its field names (allowing for inline symbol text

> as needed), and a delimited form that encodes its field names with `FlexSym` (there is no delimited form with symbol ID

Reviewer: In the sentence about delimited lists and S-expressions we explicitly called out that they end with the `0xF0` opcode. Should we mention that here too for consistency?

Author: Done. I spelled out that it ends with a `VarSym` escape (`0x00`) followed by the `0xF0` opcode.

field names).

Symbol values have link:binary-encoding.adoc#symbols_with_inline_text[two encodings]: one is the Ion 1.0-style
encoding using the symbol ID, and the other one is structurally identical to the encoding of strings, supplying its
text's UTF-8 bytes inline.

link:binary-encoding.adoc#annotations[Annotation sequences] are a prefix to the value they decorate, and no longer
have an outer length container.
They are now encoded with an opcode that specifies a single annotation with value following, an opcode that specifies
two annotations with a value following, and finally, an opcode that specifies a variable length of annotations followed
by a value.
The latter encoding is similar to how Ion 1.0 annotations are encoded with the exception that there is no
outer length.

IMPORTANT: *_Discussion_*: Should we provide an op-code for length prefixing the entire annotation?  If so, where should
it go? E.g, make the variable length SID based annotations support this.

> link:binary-encoding.adoc#integers[Integers] now use an `Int` sub-field instead of the Ion 1.0 encoding using sign-and-magnitude

Reviewer: ```suggestion
link:binary-encoding.adoc#integers[Integers] now use a `FixedInt` sub-field instead of the Ion 1.0 encoding that used sign-and-magnitude
```

(with two opcodes).

link:binary-encoding.adoc#decimals[Decimals] are structurally identical to their Ion 1.0 counterpart with the exception
of the negative zero coefficient.
The Ion 1.1 `FlexInt` encoding is two's complement, so negative zero cannot be encoded directly with it.
Instead, an encoding opcode is allocated specifically for encoding decimals with a negative zero coefficient.

link:binary-encoding.adoc#timestamps[Timestamps] no longer encode their sub-field components as octet-aligned fields.
The Ion 1.1 format uses a packed bit encoding and has a biased form (encoding the year field as an offset from 1970) to
make common encodings of timestamp easily fit in a 64-bit word for microsecond and nanosecond precision (with UTC offset
or unknown UTC offset).
Benchmarks have shown this new encoding to be 59% faster to encode and 21% faster to decode.

> A non-biased, arbitrary length timestamp with packed bit encoding is defined for cases outside the common cases.

Reviewer: ```suggestion
A non-biased, arbitrary length timestamp with packed bit encoding is defined for uncommon cases.
```

In Ion 1.0, all data is length prefixed.  While this is good for optimizing the reading of data, it requires an Ion
encoder to buffer any data in memory to calculate the data's length.  Ion 1.1 introduces optional binary syntax to allow
containers to be encoded with an end marker instead of a length prefix.

==== Low-level Binary Encoding Changes

Ion 1.0's link:https://amazon-ion.github.io/ion-docs/docs/binary.html#varuint-and-varint-fields[`VarUInt` and `VarInt`
 encoding primitives] used big-endian byte order and used the high bit of each byte to indicate whether it was the final
byte in the encoding.

> `VarInt` used an additional bit in the first byte to represent the integer's sign.

Reviewer 2: This feels like an unnecessary level of detail for this overview.  I suggest we lead by saying that 1.1 leans into little-endianness (and why) and thus replaces X and Y with P and Q.

More detailed "delta" narrative would fit better into the intro to the binary chapter.

Author: Agreed. Tracking this in #239.

<<<

> [appendix]

Author: 🗺️ An analogous table is now available in the binary encoding section, so I've removed this one rather than update it.
